### ISSUE TITLE
[low priority] [request] Always check assertions at compiletime?

### ISSUE THREAD

**Issue by awr1** [nim-lang/Nim#14950]
Are assertions *always* checked if they occur in the VM during compile time, in spite of compiler options? 

As of right now the answer is "no." If `-d:danger` or `--assertions:off` are in the compiler options, both runtime and compiletime calls to `assert` are ignored, e.g.

```nim
func foo(x :int) :int =
  assert(x < 3)
  x

const x = foo(4)
```

compiles fine.

Is there a good reason, however, to disable compiletime assertions ever, if the purpose of `-d:danger`/`--assertions:off` is really just for *runtime speed* purposes?

**Comment by juancarlospaco** [comment]
`static: doAssert` ?.

**Comment by awr1** [comment]
You can do that, sure, but that is more explicit, and maybe slightly more verbose if you want the same check to exist on debug builds at runtime (plus the programmer has to be actually thinking about separate compile time checks instead of just getting them "for free" in all cases) e.g.

```nim
when nimvm:
  doAssert(x < 3)
else:
  assert(x < 3)
```

**Comment by juancarlospaco** [comment]
Yes, you can do a term rewriting template for that and forget about it, thats on the manual, it works.

**Comment by awr1** [comment]
Why make that something for the user to have to implement though instead of just changing the behavior of `assert` to always check at compile time?

**Comment by timotheecour** [comment]
`static: doAssert x` is unrelated, and would not work here (static levels nest!).

however `assert2` below works:
```nim
template assert2*(cond: untyped, msg = ""): untyped =
  when nimvm: doAssert cond, msg
  else: assert cond, msg
```

> Why make that something for the user to have to implement though instead of just changing the behavior of assert to always check at compile time?

enabling asserts has cost for VM code too; if you want to change behavior of assert to be always on at CT, it should at least be configurable

**Comment by awr1** [comment]
> enabling asserts has cost for VM code too; if you want to change behavior of assert to be always on at CT, it should at least be configurable

Concerns about compiletime VM code optimization are a little more lax than runtime, but I agree. Maybe a compiler flag like `--assertions:vm`?

**Comment by timotheecour** [comment]
Yes
